presto-main/src/main/java/com/facebook/presto/cost/FilterStatsRule.java

```diff
@@ -43,9 +45,13 @@ public FilterStatsRule(FilterStatsCalculator filterStatsCalculator)
     }
 
     @Override
-    public Optional<PlanNodeStatsEstimate> calculate(FilterNode node, StatsProvider statsProvider, Lookup lookup, Session session, TypeProvider types)
+    public Optional<PlanNodeStatsEstimate> doCalculate(FilterNode node, StatsProvider statsProvider, Lookup lookup, Session session, TypeProvider types)
     {
         PlanNodeStatsEstimate sourceStats = statsProvider.getStats(node.getSource());
-        return Optional.of(filterStatsCalculator.filterStats(sourceStats, node.getPredicate(), session, types));
+        PlanNodeStatsEstimate estimate = filterStatsCalculator.filterStats(sourceStats, node.getPredicate(), session, types);
+        if (isDefaultFilterFactorEnabled(session) && estimate.isOutputRowCountUnknown()) {
```

Comment:
I think we should address CBO factors holistically instead of having very specialized session properties. What do you say if we create a dedicated configuration with factors, e.g:
```
public class CboConfig {
    // cost weights
    private double cpuCostWeight = 75;
    private double memoryCostWeight = 10;
    private double networkCostWeight = 15;

    // filtering coefficients
    private double joinUnknownFilterCoefficient = 0.9
    private double defaultUnknownFilterCoefficient = 0.9
    ... more coefficients here ...

    // TODO: normalization factors (derived as in https://github.com/prestodb/presto/issues/11615)
    private double joinCpuCostNormalizationFactor = ...
    private double joinMemoryCostNormalizationFactor = ...
    private double aggregationCpuCostNormalizationFactor = ...
    private double aggregationMemoryCostNormalizationFactor = ...
    ...
}
``` 

What do you think?

Reply:
Here are some of my thoughts.

For me the `0.9` coefficient doesn't have any special meaning. I'm not sure if any generic enough coefficient can be chose here. Basically the `0.9` default coefficient says "i don't know anything about filters, let's hope that if one unfiltered table is smaller than the other unfiltered table, the filtered relations will be also so".

We may want to holistically chose coefficient to the cost model (e.g.: weights for aggregation operator, where processing an aggregation symbol is more expensive than incrementing count?). But i'm not sure if by any holistic observation any better or worse default generic filter factor can be chosen.

Reply:
> But i'm not sure if by any holistic observation any better or worse default generic filter factor can be chosen.

It might not, but I wouldn't be surprised if we are consistently off (either up or down) with filter estimate. This might be due to data correlation which might differ between clusters.
We might also want to be on an educated safe side with filter estimation, e.g: +10% so that for most filters actual filter factor is lower then estimated one. Even if difference is 10% it will build up for more complex queries.

Reply:
I like that idea.

Reply:
I think it's too early to give users knobs, because they will use those knobs to achieve certain desirable effects (this is what knobs are generally for). 
But then, when we make any changes in costing (pretty any change, not necessarily as invasive as changing the meaning of "cpu cost" @ https://github.com/prestodb/presto/pull/11443#issuecomment-433982184), they won't get those effects anymore. 

In others words. 
Such knobs can be used to achieve some desirable Presto behavior at a particular version and at this version only. People will either become trapped at this version or become frustrated when they upgrade.

The model is not mature enough to expose such internals to the end users. 
We can (and we should!) have such "normalization factors" to effect useful comparisons between eg CPU cost of a project vs CPU cost of a join (today they are the same, both equal to the data size). But we should not expose any such configuration knobs yet.

Reply:
cc @martint

Reply:
I agree with @findepi's point.

Reply:
@findepi and @arhimondr you make good points.  I will leave this as is.